text/0020-block-proposer-improvements.md

```diff
@@ -0,0 +1,110 @@
+
+## Guide-level explanation
+
+[guide-level-explanation]: #guide-level-explanation
+
+### Seeing finalization before proposal
+
+The suggested solution would be to not notify the Block Proposer about proposed blocks at all. The `past` parameter passed to it would contain the hashes of deploys contained in the blocks that haven't been finalized yet, instead of the hashes of blocks themselves. This would make the `past` list a few times longer, but would remove the need for the Block Proposer to know which blocks have been proposed - it could only care about finalized blocks in this approach.
+
+One of this solution's advantages would be removing the need for detection of orphaned blocks. If we don't record proposals, we won't have to "restore" the deploys from there when a block is orphaned.
+
+Another advantage over the main alternative discussed below is the lack of a need to distinguish between identical deploy lists. We wouldn't care if multiple proto-blocks with empty lists have been proposed, we would only care about the actual set of deploys that are in the past of the block to be proposed.
+
+### Request for a block before seeing another validator's proposal
+
+We have some measures to prevent deploy replays in place: when requested for a set of deploys to propose, the Block Proposer should also receive data about the new block's past, that is, the hashes of blocks that will be in the block's past, but aren't yet finalized. The Block Proposer itself is supposed to be aware of all the finalized blocks, which will also constitute the new block's past because of what finalization means.
+
+However, the issue that can arise, is that Consensus issues a notification about finalization of a block, and a request for deploys shortly afterwards (so that it doesn't include the newly finalized block in the information about the past of the block to be proposed), but the Block Proposer receives these events in reverse order, thus making it not aware that the newly finalized block should also be considered a part of the new block's past.
+
+The solution for this issue could be to include the hash of the last finalized block in the request for deploys. The Block Proposer could then check if it already has that block as finalized, and if not, it would buffer the request and complete it when it receives the notification of the relevant block's finalization.
+
+Even with this solution in place, we could actually still experience problems. Consensus could finalize multiple blocks at once, and if the events aren't guaranteed to arrive in order, the Block Proposer could be notified of the finalization of the last block before some others. In such a case, it would still have an incomplete picture of the new block's past. To prevent this, we would have to only handle notifications about finalized blocks if the block's parent has already been finalized (and buffer finalization notifications that don't satisfy this condition, until it is satisfied).
```

Comment:
This looks to me like an unresolved question/problem. Is the implementation of this CEP addressing the issue?

Reply:
It seems like it is addressed under [**Request for a block before seeing another validator's proposal**](https://github.com/CasperLabs/ceps/pull/20/files#diff-593efcb8ed59381bcf184107760c740e5afae65aa37c9d08c487f5804fd29de4R56). It's just that the language in this paragraph makes it sounds as if it's still an open question.